### PR TITLE
ELE-142 unite tests queries - Relays on duplicate rows fix

### PR DESCRIPTION
**Pull request by IDoneShaveIt**
Unite between the queries `get_test_results`, `get_test_invocations` and `get_test_sample_data`.
This preventing a concurrency bugs, and improve the report performance.

### CODE REVIEW

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 1]
```diff
@@ -132,14 +131,14 @@ def generate_report(
...
 
-            self.execution_properties["elementary_test_count"] = len(
-                [
-                    test_metadata
-                    for test_metadata in tests_metadata
-                    if test_metadata.test_type != "dbt_test"
-                ]
-            )
-            self.execution_properties["test_result_count"] = len(tests_metadata)
+            # self.execution_properties["elementary_test_count"] = len(
```
Comment (elongl) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:134]: Accident?

File: elementary/monitor/api/tests/schema.py [hunk 2]
```diff
@@ -27,10 +27,13 @@ class DbtTestResultSchema(BaseModel):
     failed_rows_count: Optional[int] = None
 
 
-class TestMetadataSchema(BaseModel):
+class RawTestResultSchema(BaseModel):
```
Comment (elongl) [inline elementary/monitor/api/tests/schema.py:30]: Why is it called "Raw"?

File: elementary/monitor/api/tests/schema.py [hunk 3]
```diff
@@ -50,6 +53,12 @@ class TestMetadataSchema(BaseModel):
     status: str
     test_created_at: Optional[str] = None
     days_diff: float
+    invocations_order: int
+    sample_data: Optional[Union[dict, List]] = None
+
+    # pydantic has a bug with Union fields. This is how to support it.
```
Comment (elongl) [inline elementary/monitor/api/tests/schema.py:59]: This is not a bug, that's intended behavior.
I think we should remove this comment.

File: elementary/monitor/api/tests/schema.py [hunk 4]
```diff
@@ -50,6 +53,12 @@ class TestMetadataSchema(BaseModel):
     status: str
     test_created_at: Optional[str] = None
     days_diff: float
+    invocations_order: int
+    sample_data: Optional[Union[dict, List]] = None
```
Comment (elongl) [inline elementary/monitor/api/tests/schema.py:57]: Why can this be both a dict and a list?

File: elementary/monitor/api/tests/schema.py [hunk 5]
```diff
@@ -50,6 +53,12 @@ class TestMetadataSchema(BaseModel):
     status: str
     test_created_at: Optional[str] = None
     days_diff: float
+    invocations_order: int
```
Comment (elongl) [inline elementary/monitor/api/tests/schema.py:56]: I don't understand what does this represent.

File: elementary/monitor/api/tests/schema.py (hunk 2 above)
Comment (elongl) [inline elementary/monitor/api/tests/schema.py:30]: I think Pydantic models are not meant to have methods like that.
I think their main purposes should be data classes, in the sense that they hold validated data.

File: elementary/monitor/api/tests/tests.py [hunk 6]
```diff
@@ -41,35 +41,32 @@ def __init__(self, dbt_runner: DbtRunner):
...
-        tests_metadata = (
+        test_results = (
             json.loads(run_operation_response[0]) if run_operation_response else []
         )
-        tests_metadata = [
-            TestMetadataSchema(**test_metadata) for test_metadata in tests_metadata
+        test_results = [
+            RawTestResultSchema(**test_result) for test_result in test_results
         ]
         if should_cache:
```
Comment (elongl) [inline elementary/monitor/api/tests/tests.py:67]: I personally think that we should delete this cache mechanism and create a coherent data flow in the program instead.

File: elementary/monitor/api/tests/tests.py [hunk 7]
```diff
@@ -186,120 +165,118 @@ def get_test_results(
...
+                        status=raw_test_result.status,
                         affected_rows=self._parse_affected_row(
-                            results_description=test_invocation[
-                                "test_results_description"
-                            ]
+                            results_description=raw_test_result.test_results_description
                         ),
                     )
                 )
             except Exception:
```
Comment (elongl) [inline elementary/monitor/api/tests/tests.py:265]: Why do we have an `except Exception` here?

File: elementary/monitor/api/tests/tests.py [hunk 8]
```diff
@@ -341,10 +318,10 @@ def _parse_affected_row(results_description: str) -> Optional[int]:
 
     def get_total_tests_results(
         self,
-        tests_info: Optional[List[TestInfoSchema]] = None,
+        tests_metadata: Optional[List[TestMetadataSchema]] = None,
```
Comment (elongl) [inline elementary/monitor/api/tests/tests.py:321]: Why is this `Optional`?

File: elementary/monitor/dbt_project/macros/base_queries/current_tests_run_results_query.sql [hunk 9]
```diff
@@ -1,23 +1,21 @@
-{% macro current_tests_run_results_query(days_back = none, invocation_id = none) %}
+{% macro current_tests_run_results_query(days_back = none) %}
```
Comment (elongl) [inline elementary/monitor/dbt_project/macros/base_queries/current_tests_run_results_query.sql:1]: Why was this removed?
Isn't it necessary for the filters?

**Review by elongl — COMMENTED** [review]
Could you share a bit of context about the high-level logic changes presented in this PR?
I'm troubling a bit to understand how we solved the duplicates issue.
Thanks a lot.

**Comment by IDoneShaveIt** [comment]
@elongl hey!
This PR is basing on the duplicate rows fix PR.
The reason I didn't want people to review it yet, is because I didn't want to duplicate the thread about it 🙂
The changes here are only relevant for unifying the 3 queries over the elementary test reaults table, which fixes race conditions, missing keys bug and improve the performance of the report.

File: elementary/monitor/data_monitoring/data_monitoring_report.py (hunk 1 above)
Comment (IDoneShaveIt) [reply elementary/monitor/data_monitoring/data_monitoring_report.py:134]: Yes!
Will return it!

File: elementary/monitor/api/tests/schema.py (hunk 2 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/schema.py:30]: So we have 2 different schemas for test result.
One is the rows of the table elementary test results, and one is for the actual test results object we have in the report.

Can change the naming into "TestResultRowSchema" or "DbTestResultSchema" 🙂

File: elementary/monitor/api/tests/schema.py (hunk 4 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/schema.py:57]: Good question!
Currently we have a different samlpe_data structure for different test types (The UI is based on that).
This is something that we have from day 1, and we need to unify, but this isn't part of this PR 🙂

File: elementary/monitor/api/tests/schema.py (hunk 2 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/schema.py:30]: I am not sure what do you mean by methods like that?
Maybe the comment is on the wrong line?

File: elementary/monitor/dbt_project/macros/base_queries/current_tests_run_results_query.sql (hunk 9 above)
Comment (IDoneShaveIt) [reply elementary/monitor/dbt_project/macros/base_queries/current_tests_run_results_query.sql:1]: We do the filter now on the python level.
Instead of query the same data 3 times (1 for test results, 1 for invocations, 1 for data samples) we now query it only once, and filter the data when needed on the relevant method.

Invocation is not filtering the invocation part (test runs), so we need to do this filter outside of the query.

File: elementary/monitor/api/tests/schema.py (hunk 5 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/schema.py:54]: So this is part of the change in unifying the test queries.
Until now we had 3 different queries that were the same except the final filters (where clause).
We ordered the result of each test from the newest to oldest (invocations order), and we used that on the last filtering to get the latest tests runs + invocations for each test.

Now that we filter the results outside of the query (in the python), we need to return the invocations order from the query 🙂

File: elementary/monitor/api/tests/schema.py (hunk 2 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/schema.py]: I moved the more logical methods outside of the schema, so now it contains only data validation 🙂

File: elementary/monitor/api/tests/tests.py (hunk 6 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/tests.py]: Deleted!

File: elementary/monitor/api/tests/tests.py (hunk 7 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/tests.py:280]: This is here for a long time.
In case we couldn't parse an invocation, we didn't want to fail the whole report.
Now we are braver and changed our approach for breaking the report, but I am not sure if we want to remove that completely an fail on a single invocation 🤔 

My hunch is to remove it, but want to make sure first

File: elementary/monitor/dbt_project/macros/base_queries/current_tests_run_results_query.sql (hunk 9 above)
Comment (elongl) [reply elementary/monitor/dbt_project/macros/base_queries/current_tests_run_results_query.sql:1]: Awesome!

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 10]
```diff
@@ -103,22 +103,23 @@ def generate_report(
             models = self.models_api.get_models(exclude_elementary_models)
             sources = self.models_api.get_sources()
             exposures = self.models_api.get_exposures()
+
             models_runs = self.models_api.get_models_runs(
                 days_back=days_back, exclude_elementary_models=exclude_elementary_models
             )
-            tests_metadata = self.tests_api.get_tests_metadata(days_back=days_back)
 
+            test_results_db_rows = self.tests_api.get_all_test_results_db_rows(
```
Comment (elongl) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:111]: Just an idea, what do you think about having a `TestsAPI.get_tests(...)` that returns, roughly, an object that looks like this?

```py
class TestResults
    results: ...
    runs: ...
```

That way there's only a single call to `TestsAPI` to get test-related data and there's no need to re-normalize the data in `DataMonitoring`. Not very critical though.

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 11]
```diff
@@ -103,22 +103,23 @@ def generate_report(
...
+            )
             (
                 test_results,
                 test_results_totals,
                 invocation,
-            ) = self._get_test_results_and_totals(
-                days_back=days_back,
-                disable_passed_test_metrics=disable_passed_test_metrics,
-            )
+            ) = self._get_test_results_and_totals(test_results_db_rows)
```
Comment (elongl) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:120]: Can we maybe change those to return objects and not tuples?
(I know it was like that beforehand.)

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 12]
```diff
@@ -238,21 +243,18 @@ def _get_lineage(self, exclude_elementary_models: bool = False) -> LineageSchema
...
         try:
             tests_results, invocation = self.tests_api.get_test_results(
                 filter=self.filter,
-                days_back=days_back,
-                disable_passed_test_metrics=disable_passed_test_metrics,
+                test_results_db_rows=test_results_db_rows,
                 disable_samples=self.disable_samples,
             )
-            tests_info = []
+            tests_metadata = []
```
Comment (elongl) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:254]: I think it needs to be called `test_metadatas`.
The current name makes it sound like there's a single metadata for many tests.

File: elementary/monitor/api/tests/tests.py [hunk 13]
```diff
@@ -87,222 +74,227 @@ def _get_invocation_from_filter(
...
+                    display_name=test_sub_type_display_name,
+                    metrics=sample_data,
+                    result_description=test_result_db_row.test_results_description,
+                )
+            elif test_result_db_row.test_type == "schema_change":
+                test_results = ElementaryTestResultSchema(
+                    display_name=test_sub_type_display_name.lower(),
+                    result_description=test_result_db_row.test_results_description,
+                )
+        return test_results
```
Comment (elongl) [inline elementary/monitor/api/tests/tests.py:235]: image

File: elementary/monitor/api/tests/schema.py (hunk 5 above)
Comment (elongl) [reply elementary/monitor/api/tests/schema.py:54]: Does the `invocations_order` represent the `row_number` in the ordered result-set (`order by generated_at`)?
If so, why does the object itself need to be aware of its position in a list?
I think it makes more sense to pass around the ordered list when the order matters rather than store it on the object.

File: elementary/monitor/api/tests/tests.py [hunk 14]
```diff
@@ -87,222 +74,227 @@ def _get_invocation_from_filter(
...
 
+    @staticmethod
+    def _get_failed_rows_count(test_result_db_row: TestResultDBRowSchema) -> int:
+        failed_rows_count = -1
+        if (
+            test_result_db_row.status != "pass"
+            and test_result_db_row.test_results_description
+        ):
+            found_rows_number = re.search(
+                r"\d+", test_result_db_row.test_results_description
```
Comment (elongl) [inline elementary/monitor/api/tests/tests.py:198]: Why don't we use the `failures` column?

File: elementary/monitor/api/tests/tests.py [hunk 15]
```diff
@@ -10,66 +10,53 @@
...
+                disable_passed_test_metrics=disable_passed_test_metrics,
+            ),
         )
-        tests_metadata = (
+        test_results = (
             json.loads(run_operation_response[0]) if run_operation_response else []
         )
-        tests_metadata = [
-            TestMetadataSchema(**test_metadata) for test_metadata in tests_metadata
+        test_results = [
```
Comment (elongl) [inline elementary/monitor/api/tests/tests.py:56]: This should already be ordered and used rather than the `invocation_order`.

File: elementary/monitor/dbt_project/macros/get_test_results.sql [hunk 16]
```diff
@@ -1,24 +1,25 @@
...
-        tests_in_last_chosen_days as (
-            select *,
-                  {{ elementary.datediff(elementary.cast_as_timestamp('detected_at'), elementary.current_timestamp(), 'day') }} as days_diff,
-                  row_number() over (partition by model_unique_id, test_unique_id, column_name, test_sub_type order by detected_at desc) as row_number
-                from test_results
-        ),
-
-        latest_tests_in_the_last_chosen_days as (
-            select * from tests_in_last_chosen_days where row_number = 1
+        test_results_with_ivocations_order as (
```
Comment (elongl) [inline elementary/monitor/dbt_project/macros/get_test_results.sql:7]: Typo.
Should maybe be called `ordered_test_results`.

File: elementary/monitor/api/tests/tests.py [hunk 17]
```diff
@@ -87,222 +74,227 @@ def _get_invocation_from_filter(
...
-            )
-            tests_metrics = (
-                json.loads(run_operation_response[0]) if run_operation_response else {}
-            )
-
-        self.set_run_cache(key=TESTS_SAMPLE_DATA, value=tests_metrics)
-        return tests_metrics
-
-    def get_test_info_from_test_metadata(
+    def get_test_metadata_from_test_result_db_row(
```
Comment (elongl) [inline elementary/monitor/api/tests/tests.py:79]: image
image

File: elementary/monitor/api/tests/schema.py (hunk 5 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/schema.py:56]: Because we use `rank` method and not `row_number` method at the sql to order the invocations (we can have 2 different test results with 1 as the invocation order) we need to be aware of it 🙂

File: elementary/monitor/api/tests/tests.py (hunk 14 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/tests.py:202]: This is an old code when snowflake didn't return the real failures (affected rows).
So we can change it and use it, but that means that users that are using old snowflake adapter won't have a real number (only 1)

File: elementary/monitor/api/tests/tests.py (hunk 15 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/tests.py:56]: Explained on the schema why we can't relay on the order 🙂  (rank instead of row_number)

File: elementary/monitor/data_monitoring/data_monitoring_report.py (hunk 10 above)
Comment (IDoneShaveIt) [reply elementary/monitor/data_monitoring/data_monitoring_report.py:111]: Yeah I tend to agree that we need to simplify the flow 🙂 
I think that the original idea was to have an "API" call for each screen, but now we can use a single call (specially when both API's uses the same query).
I want to do it as a part of a different PR because I want to add improvements to the json as well (stop duplicating tests metadata on the json for example) which will require changes at the UI as well.

File: elementary/monitor/data_monitoring/data_monitoring_report.py (hunk 11 above)
Comment (IDoneShaveIt) [reply elementary/monitor/data_monitoring/data_monitoring_report.py:120]: Done :)

File: elementary/monitor/api/tests/schema.py (hunk 5 above)
Comment (elongl) [reply elementary/monitor/api/tests/schema.py:56]: Thanks for explaining.
Can we rename this to `invocations_rank_index`?
I think we should clarify that this represents an index within a set.

Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/schema.py:56]: Yeah sure!
